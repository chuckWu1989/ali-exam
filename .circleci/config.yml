defaults: &defaults
  working_directory: /tmp/ali-exam
  docker:
    - image: circleci/node:7.10
restore_repo: &restore_repo
  restore_cache:
    keys:
      - v1-dependencies-{{ .Branch }}-{{ .Revision }}
version: 2
jobs:
  build:
    <<: *defaults
    steps:
      - *restore_repo
      - checkout
      - run:
          name: Install dependencies
          command: npm install
      - save_cache:
          paths:
            - /tmp/ali-exam
          key: v1-dependencies-{{ .Branch }}-{{ .Revision }}
  test:
    <<: *defaults
    steps:
      - *restore_repo
      - run:
          name: Unit Testing
          command: npm test
      - run:
          name: Coverage Testing
          command: npm run test:coverage
      - run:
          name: Post Coverage Report
          command: bash <(curl -s https://codecov.io/bash)

all_tags: &all_tags
  filters:
    branches:
      only: master
workflows:
  version: 2
  build_and_test:
    jobs:
      - build:
          <<: *all_tags
      - test:
          <<: *all_tags
          requires:
            - build
