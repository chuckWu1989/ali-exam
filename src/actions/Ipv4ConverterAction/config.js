export const IPV4AVALUE = 'ipv4Value';
export const IPREGEX = /\b(\s*(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\s*\.){3}\s*(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\s*\b/;
export const SPACEREGEX = /\s/g;

export default IPV4AVALUE;
