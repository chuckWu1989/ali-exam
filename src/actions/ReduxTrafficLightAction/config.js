export const TIME = 2000;
export const EVENT = 'event';
