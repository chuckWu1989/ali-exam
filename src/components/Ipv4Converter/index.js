import React, { Component } from 'react';
import {
  IPV4TITLE,
  INTTITLE,
  TRANSFERBTNTITLE,
  INTVALUE,
  ERROR,
  ONINITIAL,
  ONDISPOSE,
  GETIPV4STRING,
  TRANSFER,
  defaultProps,
  propTypes,
} from './props';
import './style.less';

class Ipv4Converter extends Component {
  componentWillMount() {
    this.props[ONINITIAL]();
  }
  componentWillUnmount() {
    this.props[ONDISPOSE]();
  }
  render() {
    const {
      [IPV4TITLE]: ipv4Title,
      [INTTITLE]: intTitle,
      [TRANSFERBTNTITLE]: transferBtnTitle,
      [GETIPV4STRING]: getIpv4String,
      [ERROR]: error,
      [TRANSFER]: transfer,
      [INTVALUE]: intValue,
    } = this.props;
    return (
      <div className="ipv4-converter-style">
        <div>
          <div className="ipv4-title">
            {ipv4Title}
          </div>
          <div className="ipv4-input">
            <input type="text" onChange={getIpv4String} />
          </div>
        </div>
        <div className="btn-area">
          <input type="button" value={transferBtnTitle} onClick={transfer} />
        </div>
        <div>
          <div className="ipv4-title">
            {intTitle}
          </div>
          <div className="ipv4-input">
            <input type="text" value={intValue} />
            <div className="error">{error}</div>
          </div>
        </div>
      </div>
    );
  }
}
Ipv4Converter.defaultProps = defaultProps;
Ipv4Converter.propTypes = propTypes;

export default Ipv4Converter;
