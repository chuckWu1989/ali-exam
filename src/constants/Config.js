export const SSTORE = 'store';
export const SDATA = 'data';
export const SPROPS = 'props';
export const SCHILDREN = 'children';
export const SUICONTROL = 'uiControl';
export const SEVENT = 'event';
