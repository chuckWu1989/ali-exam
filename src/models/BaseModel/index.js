import Immutable from 'immutable';
import {
  SDATA,
  SPROPS,
  SCHILDREN,
  SUICONTROL,
  SEVENT,
} from '../../constants/Config';

export default Immutable.fromJS({
  [SDATA]: {},
  [SPROPS]: undefined,
  [SCHILDREN]: undefined,
  [SUICONTROL]: {},
  [SEVENT]: {},
});
