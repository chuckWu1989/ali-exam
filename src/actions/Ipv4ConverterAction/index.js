import { NAME, INTVALUE, ERRORMSG, ERROR } from '../../components/Ipv4Converter/props';
import { IPV4AVALUE, IPREGEX, SPACEREGEX } from './config';

export const ipv4ToInt = (ipv4, err) => {
  let errorMsg;
  let intValue;
  const regex = new RegExp(IPREGEX);
  if (!regex.test(ipv4)) {
    errorMsg = err;
  } else {
    intValue = ipv4.split('.').map((item, idx, array) => (
      parseInt(item.replace(SPACEREGEX, ''), 10) * (256 ** (array.length - idx - 1))
    )).reduce((prev, curr) => prev + curr);
  }
  return { intValue, errorMsg };
};
export const transfer = props => (
  (dispatch, getState, tools) => {
    const { [NAME]: name, [ERRORMSG]: err } = props;
    const { getData, setData } = tools;
    let payload = getData(name);
    const ipv4Value = payload.get(IPV4AVALUE);
    const { intValue, errorMsg } = ipv4ToInt(ipv4Value, err);
    payload = payload.set(INTVALUE, String(intValue)).set(ERROR, errorMsg);
    setData(name, payload);
  }
);
export const getIpv4String = (event, props) => (
  (dispatch, getState, tools) => {
    const { [NAME]: name } = props;
    const { target: { value } } = event;
    const { getData, setData } = tools;
    let payload = getData(name);
    payload = payload.set(IPV4AVALUE, value);
    setData(name, payload);
  }
);
export const onInitial = props => (
  (dispatch, getState, tools) => {
    const { [NAME]: name } = props;
    const { superInit } = tools;
    superInit(name, props);
  }
);
export const onDispose = props => (
  (dispatch, getState, tools) => {
    const { [NAME]: name } = props;
    const { superDispose } = tools;
    superDispose(name);
  }
);
