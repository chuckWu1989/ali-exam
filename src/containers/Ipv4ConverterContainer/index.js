import connect from '../../hocs/connect';
import Ipv4Converter from '../../components/Ipv4Converter';
import { ONINITIAL, ONDISPOSE, GETIPV4STRING, TRANSFER, INTVALUE, ERROR, NAME, defaultProps } from '../../components/Ipv4Converter/props';
import { onInitial, onDispose, getIpv4String, transfer } from '../../actions/Ipv4ConverterAction';
import { SSTORE, SDATA } from '../../constants/Config';

export const mapStateToProps = (state, props) => ({
  [INTVALUE]: state.getIn([SSTORE, props[NAME], SDATA, INTVALUE]),
  [ERROR]: state.getIn([SSTORE, props[NAME], SDATA, ERROR]),
});
export const mapDispatchToProps = (dispatch, props) => ({
  [ONINITIAL]: () => dispatch(onInitial(props)),
  [GETIPV4STRING]: event => dispatch(getIpv4String(event, props)),
  [TRANSFER]: () => dispatch(transfer(props)),
  [ONDISPOSE]: () => dispatch(onDispose(props)),
});
export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Ipv4Converter, defaultProps);
