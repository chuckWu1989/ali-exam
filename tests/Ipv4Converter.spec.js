import React from 'react';
import toJson from 'enzyme-to-json';
import configureStore from 'redux-mock-store';
import Immutable from 'immutable';
import {
  defaultProps,
  NAME,
  IPV4TITLE,
  INTTITLE,
  INTVALUE,
  ERROR,
  ONINITIAL,
  ONDISPOSE,
  GETIPV4STRING,
  TRANSFER,
  ERRORMSG,
} from '../src/components/Ipv4Converter/props';
import { IPV4AVALUE } from '../src/actions/Ipv4ConverterAction/config';
import Ipv4Converter from '../src/components/Ipv4Converter';
import { mapStateToProps, mapDispatchToProps } from '../src/containers/Ipv4ConverterContainer';
import customThunk from '../src/store/customThunk';
import { SSTORE, SDATA } from '../src/constants/Config';
import BaseModel from '../src/models/BaseModel';
import * as actions from '../src/actions/Ipv4ConverterAction';

const mockStore = configureStore([customThunk]);
const { shallow } = global;
const props = {
  [NAME]: 'test',
  [IPV4TITLE]: 'IPV4 test title',
  [INTTITLE]: 'Int test title',
  [INTVALUE]: 'test value',
  [ERRORMSG]: 'Wrong',
};

describe('Ipv4Converter', () => {
  describe('component', () => {
    it('should be rendered properly', () => {
      const ownProps = {
        ...props,
      };
      const tree = toJson(shallow(<Ipv4Converter {...ownProps} />));
      expect(tree).toMatchSnapshot();
    });
    it('should call methods during lifecycle', () => {
      const ownProps = {
        ...props,
        [ONINITIAL]: jest.fn(),
        [ONDISPOSE]: jest.fn(),
      };
      const wrapper = shallow(
        <Ipv4Converter {...ownProps} />,
        { lifecycleExperimental: true },
      );
      wrapper.unmount();
      expect(ownProps[ONINITIAL]).toHaveBeenCalledTimes(1);
      expect(ownProps[ONDISPOSE]).toHaveBeenCalledTimes(1);
    });
  });
  describe('container', () => {
    const state = {
      getIn: () => {},
    };
    const dispatch = () => {};
    const mapState = mapStateToProps(state, props);
    const mapDispatch = mapDispatchToProps(dispatch, props);
    describe('mapStateToProps', () => {
      it('should pass intValue', () => {
        expect(INTVALUE in mapState).toBeTruthy();
      });
      it('should pass error', () => {
        expect(ERROR in mapState).toBeTruthy();
      });
    });
    describe('mapDispatchToProps', () => {
      it('should pass onInitial', () => {
        expect(ONINITIAL in mapDispatch).toBeTruthy();
      });
      it('should pass getIpv4String', () => {
        expect(GETIPV4STRING in mapDispatch).toBeTruthy();
      });
      it('should pass transfer', () => {
        expect(TRANSFER in mapDispatch).toBeTruthy();
      });
      it('should pass onDispose', () => {
        expect(ONDISPOSE in mapDispatch).toBeTruthy();
      });
    });
  });
  describe('action', () => {
    let store;
    beforeEach(() => {
      store = mockStore(Immutable.Map({}).setIn([SSTORE, props[NAME]], BaseModel));
    });
    describe('onInitial', () => {
      it('should call superInit', () => {
        store.dispatch(actions.onInitial(defaultProps));
        expect(store.getActions().length).toBe(2);
      });
    });
    describe('onDispose', () => {
      it('should call superDispose', () => {
        store.dispatch(actions.onDispose(defaultProps));
        expect(store.getActions().length).toBe(1);
      });
    });
    describe('getIpv4String', () => {
      it('should get ipv4 string in payload', () => {
        const ownProps = { ...defaultProps, ...props };
        const target = '123';
        const event = { target: { value: target } };
        store.dispatch(actions.getIpv4String(event, ownProps));
        expect(store.getActions()[0].payload.payload.get(IPV4AVALUE)).toBe(target);
      });
    });
    describe('ipv4ToInt', () => {
      it('should return errorMsg if the format of ipv4 string is invalid', () => {
        const ipv4 = 'Error format';
        const err = 'Error message';
        const { intValue, errorMsg } = actions.ipv4ToInt(ipv4, err);
        expect(errorMsg).toBe(err);
        expect(intValue).toBeUndefined();
      });
      it('should return int value if the format of ipv4 string is valid', () => {
        const ipv4 = '182 .234. 35 .42';
        const err = 'Error message';
        const { intValue, errorMsg } = actions.ipv4ToInt(ipv4, err);
        expect(errorMsg).toBeUndefined();
        expect(intValue).toBe(3068797738);
      });
    });
    describe('transfer', () => {
      it('should emit payload with transfered data', () => {
        const ownProps = { ...defaultProps, ...props };
        const ipv4 = '182 .234. 35 .42';
        const newModel = BaseModel.setIn([SDATA, IPV4AVALUE], ipv4);
        store = mockStore(Immutable.Map({}).setIn([SSTORE, props[NAME]], newModel));
        store.dispatch(actions.transfer(ownProps));
        expect(store.getActions()[0].payload.payload.get(INTVALUE)).toBe('3068797738');
      });
    });
  });
});
