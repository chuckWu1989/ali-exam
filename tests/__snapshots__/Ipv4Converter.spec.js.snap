// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ipv4Converter component should be rendered properly 1`] = `
<div
  className="ipv4-converter-style"
>
  <div
    className="input-area"
  >
    <div
      className="ipv4-title"
    >
      IPV4 test title
    </div>
    <div
      className="ipv4-input"
    >
      <input
        onChange={[Function]}
        type="text"
      />
    </div>
  </div>
  <div
    className="btn-area"
  >
    <input
      onClick={[Function]}
      type="button"
      value="Transfer"
    />
  </div>
  <div
    className="input-area"
  >
    <div
      className="ipv4-title"
    >
      Int test title
    </div>
    <div
      className="ipv4-input"
    >
      <input
        readOnly={true}
        type="text"
        value="test value"
      />
      <div
        className="error"
      />
    </div>
  </div>
</div>
`;
